File: internal/core/domain/entities/payment.go
package entities

import (
	"time"

	"github.com/google/uuid"
)

const (
	DEFAULT_PAYMENT_EXPIRES_IN_MINUTES = 30
)

type PaymentStatus string

const (
	PaymentStatus_Pending  = "pending"
	PaymentStatus_Paid     = "paid"
	PaymentStatus_Rejected = "rejected"
	PaymentStatus_Canceled = "canceled"
)

type PaymentMethod string

const (
	PaymentMethod_Pix = "pix"
)

type Payment struct {
	Id          string        `dynamodbav:"id"`
	ExternalId  string        `dynamodbav:"external_id"`
	Status      PaymentStatus `dynamodbav:"status"`
	Method      PaymentMethod `dynamodbav:"method"`
	Description string        `dynamodbav:"description"`
	Code        string        `dynamodbav:"code"`
	Value       float64       `dynamodbav:"value"`
	ExpiresAt   int64         `dynamodbav:"expires_at"`
	CreatedAt   int64         `dynamodbav:"created_at"`
	UpdatedAt   int64         `dynamodbav:"updated_at"`
}

func NewPayment(externalId string, value float64, description string, code string) *Payment {
	return &Payment{
		Id:          uuid.New().String(),
		ExternalId:  externalId,
		Status:      PaymentStatus_Pending,
		Method:      PaymentMethod_Pix,
		Description: description,
		Value:       value,
		Code:        code,
		ExpiresAt:   time.Now().Add(DEFAULT_PAYMENT_EXPIRES_IN_MINUTES * time.Minute).Unix(),
		CreatedAt:   time.Now().Unix(),
		UpdatedAt:   time.Now().Unix(),
	}
}

File: src/core/domain/entities/payment.ts
import { randomUUID } from "node:crypto"
import { PaymentMethod, PaymentStatus } from "./payment.types"

export const DEFAULT_PAYMENT_EXPIRES_IN_SECONDS = 30 * 60

export class Payment {
	id: string
	external_id: string
	status: PaymentStatus
	method: PaymentMethod
	description: string
	code: string
	value: number
	expires_at: number
	created_at: number
	updated_at: number

	constructor({
		id,
		externalId,
		status,
		method,
		description,
		value,
		code,
		expiresAt,
		createdAt,
		updatedAt,
	}: {
		id: string,
		externalId: string,
		status: PaymentStatus,
		method: PaymentMethod,
		description: string,
		value: number,
		code: string,
		expiresAt: number,
		createdAt: number,
		updatedAt: number,
	}) {
		this.id = id
		this.external_id = externalId
		this.status = status
		this.method = method
		this.description = description
		this.value = value
		this.code = code
		this.expires_at = expiresAt
		this.created_at = createdAt
		this.updated_at = updatedAt
	}

	static create(externalId: string, value: number, description: string, code: string): Payment {
		return new Payment({
			id: randomUUID(),
			externalId,
			status: PaymentStatus.Pending,
			method: PaymentMethod.Pix,
			description,
			value,
			code,
			expiresAt: Date.now() + DEFAULT_PAYMENT_EXPIRES_IN_SECONDS,
			createdAt: Date.now(),
			updatedAt: Date.now(),
		})
	}
}
